This patch adds support for GETTEXTIOENCODING environment variable that can be
set to override console codepage autodetection. This is particularly useful
when redirecting the output to a pipe (console CP detection fails) and using
localized messages. Inspired by Python's PYTHONIOENCODING.

WARNING!

This is considered deprecated. Use OUTPUT_CHARSET variable instead.
See https://lists.gnu.org/archive/html/bug-gettext/2019-04/msg00055.html


--- a/gettext-runtime/gnulib-lib/localcharset.c
+++ b/gettext-runtime/gnulib-lib/localcharset.c
@@ -688,6 +688,12 @@ locale_charset (void)
 {
   const char *codeset;
 
+  /* Force output encoding, particularly useful for redirected output on win32
+     where charset detection fails when no actual console is attached. */
+  const char *forced_encoding = getenv("GETTEXTIOENCODING");
+  if (forced_encoding)
+    return forced_encoding;
+
 #if HAVE_LANGINFO_CODESET || defined WINDOWS_NATIVE || defined OS2
 
 # if HAVE_LANGINFO_CODESET

--- a/gettext-runtime/intl/localcharset.c
+++ b/gettext-runtime/intl/localcharset.c
@@ -688,6 +688,12 @@ locale_charset (void)
 {
   const char *codeset;
 
+  /* Force output encoding, particularly useful for redirected output on win32
+     where charset detection fails when no actual console is attached. */
+  const char *forced_encoding = getenv("GETTEXTIOENCODING");
+  if (forced_encoding)
+    return forced_encoding;
+
 #if HAVE_LANGINFO_CODESET || defined WINDOWS_NATIVE || defined OS2
 
 # if HAVE_LANGINFO_CODESET

--- a/gettext-tools/gnulib-lib/localcharset.c
+++ b/gettext-tools/gnulib-lib/localcharset.c
@@ -688,6 +688,12 @@ locale_charset (void)
 {
   const char *codeset;
 
+  /* Force output encoding, particularly useful for redirected output on win32
+     where charset detection fails when no actual console is attached. */
+  const char *forced_encoding = getenv("GETTEXTIOENCODING");
+  if (forced_encoding)
+    return forced_encoding;
+
 #if HAVE_LANGINFO_CODESET || defined WINDOWS_NATIVE || defined OS2
 
 # if HAVE_LANGINFO_CODESET

--- a/gettext-tools/libgettextpo/localcharset.c
+++ b/gettext-tools/libgettextpo/localcharset.c
@@ -688,6 +688,12 @@ locale_charset (void)
 {
   const char *codeset;
 
+  /* Force output encoding, particularly useful for redirected output on win32
+     where charset detection fails when no actual console is attached. */
+  const char *forced_encoding = getenv("GETTEXTIOENCODING");
+  if (forced_encoding)
+    return forced_encoding;
+
 #if HAVE_LANGINFO_CODESET || defined WINDOWS_NATIVE || defined OS2
 
 # if HAVE_LANGINFO_CODESET
